ASSERTION FAILED: !retainedPointer || !m_controlBlock->objectHasBeenDeleted() in the GPUProcess

cdumez opened the pull request:
#### 2cf73e50c05c8cfa3980090d0d635e8a67611d63
<pre>
ASSERTION FAILED: !retainedPointer || !m_controlBlock-&gt;objectHasBeenDeleted() in the GPUProcess
<a href="https://bugs.webkit.org/show_bug.cgi?id=254186">https://bugs.webkit.org/show_bug.cgi?id=254186</a>
rdar://106963694

Reviewed by Alex Christensen.

The assertion was being hit on a background thread inside MediaRecorderPrivateWriter::compressedVideoOutputBufferCallback()
when constructing a ThreadSafeWeakPtr from MediaRecorderPrivateWriter pointer. You can see in the crash logs that the
MediaRecorderPrivateWriter destructor is currently running on the main thread at the time.

MediaRecorderPrivateWriter owns the VideoSampleBufferCompressor &amp; AudioSampleBufferCompressor via m_videoCompressor &amp;
m_audioCompressor unique pointers. When constructed, VideoSampleBufferCompressor &amp; AudioSampleBufferCompressor
register triggers that will cause compressedVideoOutputBufferCallback() &amp; compressedAudioOutputBufferCallback() to
get called off the main thread with a pointer to the MediaRecorderPrivateWriter object.

I updated the VideoSampleBufferCompressor &amp; AudioSampleBufferCompressor destructors to unregister those triggers so
that we guarantee that the compressedVideoOutputBufferCallback() &amp; compressedAudioOutputBufferCallback() functions
cannot get called once the VideoSampleBufferCompressor &amp; AudioSampleBufferCompressor objects have been destroyed.
Given that these objects are owned by the MediaRecorderPrivateWriter via unique_ptr, it also guarantees that the
MediaRecorderPrivateWriter pointer passed to these callbacks will never be stale / dead.

However, even with this fix, we&apos;d still hit the assertion in ThreadSafeWeakPtr, even though the code is now safe.
The reason is that `m_controlBlock-&gt;objectHasBeenDeleted()` return true as soon as the object has started deletion.
It doesn&apos;t indicate that the object has been deleted. I have renamed this function to objectHasStartedDeletion()
for clarity.

I am also removing these assertions from ThreadSafeWeakPtr because there are too restrictive and would prevent this
safe code from using ThreadSafeWeakPtr. At the time, compressedVideoOutputBufferCallback() constructs the
TheadSafeWeakPtr, the MediaRecorderPrivateWriter destructor may have started running but we know the objects is
still alive thanks to the trigger removal in the VideoSampleBufferCompressor destructor. The callback just needs a
ThreadSafeWeakPtr to check if the object is STILL alive once the dispatch to the main thread has occurred. This is
important since the MediaRecorderPrivateWriter gets destroyed on the main thread and since the object may have been
destroyed by the time the task runs on the main thread. ThreadSafeWeakPtr is able to deal with this just fine once
you remove the `objectHasStartedDeletion` assertions. It will successfully create the ThreadSafeWeakPtr when the
object has started destruction. If you later call `get()` on this ThreadSafeWeakPtr, it will return null, as
expected.

* Source/WTF/wtf/ThreadSafeWeakHashSet.h:
* Source/WTF/wtf/ThreadSafeWeakPtr.h:
(WTF::ThreadSafeWeakPtrControlBlock::objectHasStartedDeletion const):
(WTF::ThreadSafeWeakPtr::ThreadSafeWeakPtr):
(WTF::ThreadSafeWeakPtr::operator=):
(WTF::ThreadSafeWeakPtrControlBlock::objectHasBeenDeleted const): Deleted.
* Source/WebCore/PAL/pal/cf/CoreMediaSoftLink.cpp:
* Source/WebCore/PAL/pal/cf/CoreMediaSoftLink.h:
* Source/WebCore/platform/mediarecorder/cocoa/AudioSampleBufferCompressor.h:
* Source/WebCore/platform/mediarecorder/cocoa/AudioSampleBufferCompressor.mm:
(WebCore::AudioSampleBufferCompressor::~AudioSampleBufferCompressor):
(WebCore::AudioSampleBufferCompressor::initialize):
* Source/WebCore/platform/mediarecorder/cocoa/MediaRecorderPrivateWriterCocoa.mm:
(WebCore::MediaRecorderPrivateWriter::~MediaRecorderPrivateWriter):
* Source/WebCore/platform/mediarecorder/cocoa/VideoSampleBufferCompressor.h:
* Source/WebCore/platform/mediarecorder/cocoa/VideoSampleBufferCompressor.mm:
(WebCore::VideoSampleBufferCompressor::~VideoSampleBufferCompressor):
(WebCore::VideoSampleBufferCompressor::initialize):

Canonical link: <a href="https://commits.webkit.org/265268@main">https://commits.webkit.org/265268@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/de172f4bd499e3908939676f507a4dbd739857d7

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/10430 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/10660 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/10935 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/12070 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/10002 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/10442 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/12779 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/10614 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/12966 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/10589 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/11388 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/8784 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/12468 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/8692 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/9441 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/16700 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/8806 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/9783 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/9593 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/12840 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/9885 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/10025 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/8132 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/10557 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/9183 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/2756 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/2505 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/13438 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/10847 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/9886 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/2679 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

achristensen07 reviewed (approved):
Yeah, we're going to need to be careful with the ThreadSafeWeakPtr ctors and operator='s that pass in raw pointers and references.  This is fine.  In the future, we may consider removing them and only allowing you to interact with ThreadSafeWeakPtr through Ref and RefPtr.

cdumez committed:
ASSERTION FAILED: !retainedPointer || !m_controlBlock->objectHasBeenDeleted() in the GPUProcess
https://bugs.webkit.org/show_bug.cgi?id=254186
rdar://106963694

Reviewed by Alex Christensen.

The assertion was being hit on a background thread inside MediaRecorderPrivateWriter::compressedVideoOutputBufferCallback()
when constructing a ThreadSafeWeakPtr from MediaRecorderPrivateWriter pointer. You can see in the crash logs that the
MediaRecorderPrivateWriter destructor is currently running on the main thread at the time.

MediaRecorderPrivateWriter owns the VideoSampleBufferCompressor & AudioSampleBufferCompressor via m_videoCompressor &
m_audioCompressor unique pointers. When constructed, VideoSampleBufferCompressor & AudioSampleBufferCompressor
register triggers that will cause compressedVideoOutputBufferCallback() & compressedAudioOutputBufferCallback() to
get called off the main thread with a pointer to the MediaRecorderPrivateWriter object.

I updated the VideoSampleBufferCompressor & AudioSampleBufferCompressor destructors to unregister those triggers so
that we guarantee that the compressedVideoOutputBufferCallback() & compressedAudioOutputBufferCallback() functions
cannot get called once the VideoSampleBufferCompressor & AudioSampleBufferCompressor objects have been destroyed.
Given that these objects are owned by the MediaRecorderPrivateWriter via unique_ptr, it also guarantees that the
MediaRecorderPrivateWriter pointer passed to these callbacks will never be stale / dead.

However, even with this fix, we'd still hit the assertion in ThreadSafeWeakPtr, even though the code is now safe.
The reason is that `m_controlBlock->objectHasBeenDeleted()` return true as soon as the object has started deletion.
It doesn't indicate that the object has been deleted. I have renamed this function to objectHasStartedDeletion()
for clarity.

I am also removing these assertions from ThreadSafeWeakPtr because there are too restrictive and would prevent this
safe code from using ThreadSafeWeakPtr. At the time, compressedVideoOutputBufferCallback() constructs the
TheadSafeWeakPtr, the MediaRecorderPrivateWriter destructor may have started running but we know the objects is
still alive thanks to the trigger removal in the VideoSampleBufferCompressor destructor. The callback just needs a
ThreadSafeWeakPtr to check if the object is STILL alive once the dispatch to the main thread has occurred. This is
important since the MediaRecorderPrivateWriter gets destroyed on the main thread and since the object may have been
destroyed by the time the task runs on the main thread. ThreadSafeWeakPtr is able to deal with this just fine once
you remove the `objectHasStartedDeletion` assertions. It will successfully create the ThreadSafeWeakPtr when the
object has started destruction. If you later call `get()` on this ThreadSafeWeakPtr, it will return null, as
expected.

* Source/WTF/wtf/ThreadSafeWeakHashSet.h:
* Source/WTF/wtf/ThreadSafeWeakPtr.h:
(WTF::ThreadSafeWeakPtrControlBlock::objectHasStartedDeletion const):
(WTF::ThreadSafeWeakPtr::ThreadSafeWeakPtr):
(WTF::ThreadSafeWeakPtr::operator=):
(WTF::ThreadSafeWeakPtrControlBlock::objectHasBeenDeleted const): Deleted.
* Source/WebCore/PAL/pal/cf/CoreMediaSoftLink.cpp:
* Source/WebCore/PAL/pal/cf/CoreMediaSoftLink.h:
* Source/WebCore/platform/mediarecorder/cocoa/AudioSampleBufferCompressor.h:
* Source/WebCore/platform/mediarecorder/cocoa/AudioSampleBufferCompressor.mm:
(WebCore::AudioSampleBufferCompressor::~AudioSampleBufferCompressor):
(WebCore::AudioSampleBufferCompressor::initialize):
* Source/WebCore/platform/mediarecorder/cocoa/MediaRecorderPrivateWriterCocoa.mm:
(WebCore::MediaRecorderPrivateWriter::~MediaRecorderPrivateWriter):
* Source/WebCore/platform/mediarecorder/cocoa/VideoSampleBufferCompressor.h:
* Source/WebCore/platform/mediarecorder/cocoa/VideoSampleBufferCompressor.mm:
(WebCore::VideoSampleBufferCompressor::~VideoSampleBufferCompressor):
(WebCore::VideoSampleBufferCompressor::initialize):

Canonical link: https://commits.webkit.org/265268@main
